File: Python/Lecture_01_Lab_Exercise_2_Integration.py
'''
========================================================================
        ╦ ╦┌─┐┌─┐┌─┐┌─┐┌┬┐  ╔╦╗┌─┐┌─┐┌┬┐┬ ┬  ╔╗ ┌─┐┬  ┌─┐┬ ┬┌─┐
        ╠═╣│ │└─┐└─┐├─┤│││  ║║║├─┤│ ┬ ││└┬┘  ╠╩╗├─┤│  ├─┤├─┤├─┤
        ╩ ╩└─┘└─┘└─┘┴ ┴┴ ┴  ╩ ╩┴ ┴└─┘─┴┘ ┴   ╚═╝┴ ┴┴─┘┴ ┴┴ ┴┴ ┴
========================================================================
# Author: Hossam Magdy Balaha
# Permissions and Citation: Refer to the README file.
'''

# This script computes a definite integral numerically and visualizes the area under
# a Gaussian curve for instructional purposes.

# Import necessary libraries.
import numpy as np  # Library for numerical operations.
import matplotlib.pyplot as plt  # Library for creating plots.
from scipy.integrate import quad  # Function for numerical integration.


# Define the function f(x) = e^(-x^2) as the integrand for the numerical integration example.
def f(x):
  return np.exp(-x ** 2)  # Return the value of the Gaussian function at x.


# Compute the definite integral of f(x) from 0 to 1 using quad() to demonstrate numerical integration accuracy.
result, _ = quad(f, 0, 1)  # Use quad() to compute the integral and store the result.
print("Integral from 0 to 1:", result)  # Print the computed integral value.

# Generate x values for plotting over the range [0, 1] with 500 points for smooth visualization.
xVals = np.linspace(0, 1, 500)  # Create an array of 500 evenly spaced values between 0 and 1.

# Evaluate the function f(x) at each point in xVals to obtain y-values for plotting.
yVals = f(xVals)  # Compute the y-values of the function f(x) for visualization.

# Plot the function and shade the area under the curve to illustrate the definite integral visually.
plt.figure(figsize=(10, 6))  # Create a new figure with a specified size.
plt.plot(xVals, yVals, label="f(x) = e^(-x^2)", color="blue")  # Plot f(x) with a label and blue color.

# Shade the area under the curve between x = 0 and x = 1 for emphasis.
plt.fill_between(xVals, yVals, color="lightblue", alpha=0.5)  # Fill the area under the curve with light blue color.

# Add reference lines at x=0 and y=0 to mark integral limits and baseline.
plt.axhline(0, color="black", linewidth=0.8, linestyle="--")  # Add a horizontal line at y=0.
plt.axvline(0, color="black", linewidth=0.8, linestyle="--")  # Add a vertical line at x=0.
plt.axvline(1, color="black", linewidth=0.8, linestyle="--")  # Add a vertical line at x=1.

# Add title, axis labels, grid, and legend for clarity in the plotted figure.
plt.title("Area Under the Curve: Integral of f(x) = e^(-x^2) from 0 to 1.")  # Add a title to the plot.
plt.xlabel("x")  # Label the x-axis.
plt.ylabel("f(x)")  # Label the y-axis.
plt.grid()  # Add a grid to the plot for better readability.
plt.legend()  # Add a legend to identify the curve.

# Save the plot as a PNG file with high resolution for inclusion in documents or slides.
plt.savefig("Lecture_01_Lab_Exercise_2_Integration.png", dpi=300, bbox_inches="tight")
plt.show()  # Display the plot.

File: MATLAB/Lecture_01_Lab_Exercise_2_Integration.m
% ===========================================================================
%         ╦ ╦┌─┐┌─┐┌─┐┌─┐┌┬┐  ╔╦╗┌─┐┌─┐┌┬┐┬ ┬  ╔╗ ┌─┐┬  ┌─┐┬ ┬┌─┐
%         ╠═╣│ │└─┐└─┐├─┤│││  ║║║├─┤│ ┬ ││└┬┘  ╠╩╗├─┤│  ├─┤├─┤├─┤
%         ╩ ╩└─┘└─┘└─┘┴ ┴┴ ┴  ╩ ╩┴ ┴└─┘─┴┘ ┴   ╚═╝┴ ┴┴─┘┴ ┴┴ ┴┴ ┴
% ===========================================================================
%
% Author: Hossam Magdy Balaha
% Permissions and Citation: Refer to the README file.

% This script computes a definite integral numerically and visualizes the area under
% a Gaussian curve for instructional purposes.

% Import necessary toolboxes (Symbolic Math Toolbox is optional for symbolic integration).
syms x;  % Define the symbolic variable 'x' if needed.

% Define the function f(x) = e^(-x^2) as an anonymous function for numerical evaluation.
f = @(x) exp(-x.^2);  % Anonymous function for f(x).

% Compute the definite integral of f(x) from 0 to 1 using integral() to demonstrate numerical integration.
result = integral(f, 0, 1);  % Use integral() to compute the definite integral.
disp(['Integral from 0 to 1: ', num2str(result)]);  % Display the computed integral value.

% Generate x values for plotting over the range [0, 1] with 500 points for smooth visualization.
xVals = linspace(0, 1, 500)';  % Create an array of 500 evenly spaced values between 0 and 1.

% Evaluate the function f(x) at each point in xVals to obtain y-values for plotting.
yVals = f(xVals);  % Compute the y-values of the function f(x) for visualization.

% Plot the function and shade the area under the curve to illustrate the definite integral visually.
figure;  % Create a new figure window.
plot(xVals, yVals, 'b', 'LineWidth', 1.5);
hold on;  % Plot f(x) in blue with a line width of 1.5.

% Shade the area under the curve between x = 0 and x = 1 for emphasis using the fill function.
xFill = [0; xVals; 1];  % Concatenate x-values: start at 0, go through xVals, and end at 1.
yFill = [0; yVals; 0];  % Concatenate y-values: start at 0, go through yVals, and return to 0.
fill(xFill, yFill, 'blue', 'FaceAlpha', 0.5);  % Fill the area under the curve.

% Add reference lines at x=0 and y=0 to mark integral limits and baseline for the plot.
line([0, 0], get(gca, 'YLim'), 'Color', 'black', 'LineWidth', 0.8, 'LineStyle', '--');  % Vertical line at x=0.
line(get(gca, 'XLim'), [0, 0], 'Color', 'black', 'LineWidth', 0.8, 'LineStyle', '--');  % Horizontal line at y=0.
line([1, 1], get(gca, 'YLim'), 'Color', 'black', 'LineWidth', 0.8, 'LineStyle', '--');  % Vertical line at x=1.

% Add title, axis labels, grid, and legend to the plot for clarity and presentation.
title('Area Under the Curve: Integral of f(x) = e^{-x^2} from 0 to 1');
xlabel('x');
ylabel('f(x)');
grid on;
legend('f(x) = e^{-x^2}', 'Location', 'Best');

hold off;  % Release the plot hold.
